### ISSUE TITLE
Bayesian Optimization

### ISSUE THREAD

**Issue by ngoodman** [uber/pyro#836]
We should make a module on top of pyro for bayesian optimization. (This could initially live in contrib.)

Now that we have working GPs it should be pretty straightforward to implement standard BO. We could also do BO based on adaptive basis function regression as the function distribution.

@ysaatchi @fehiepsi (etal) please edit to add details as you see fit!

**Comment by fehiepsi** [comment]
Yeah, this is an interesting direction. I would like to work on this (might be two weeks later) after implementing Duo Averaging for HMC. It would be great if someone takes an initial step. :)

**Comment by eb8680** [comment]
There are quite a few GP features that we should add first - more documentation/examples/tutorials, batching, NUTS with dual averaging for hyperparameters, a simpler interface for learning, etc.  I'd recommend focusing on those at least until after the Pyro 0.2.0 release.

**Comment by ysaatchi** [comment]
This is an awesome idea and definitely lives on top of pyro as @ngoodman said! 
One challenge is optimizing acquisition functions with derivatives. This requires computing derivatives w.r.t. GP inputs. @fehiepsi would this be straightforward? 
I also agree with @eb8680 RE priorities. I think we would also like some additional likelihoods before Pyro 0.2.0, especially for classification.

**Comment by tobyclh** [comment]
Hello, I am here to share about what I learn so far trying to do BO with pyro :)
regarding derivatives w.r.t. to GP inputs, theoretically derivatives of GP can be written as they do [here](http://mlg.eng.cam.ac.uk/mchutchon/DifferentiatingGPs.pdf), but I haven't tested it myself.
And for the acquisition functions, folks in [Robo](https://github.com/automl/RoBO/tree/master/robo/acquisition_functions) did a great job implementing the most common ones so I assume that will be a good start.

**Comment by fehiepsi** [comment]
@ysaatchi @tobyclh It seems to me that computing derivatives is quite straightforward (just make the input variables requires_grad when we need to compute). Let me some time to catch up with literature before I can make any claim.

For more features of GP, I made a tracking list at #832. Will start to implement them after duo averaging stuff.

**Comment by tobyclh** [comment]
I have reviewed some of frameworks for API design idea, including
[AutoML's RoBO](https://github.com/automl/RoBO) (a top level BO object with input string to specify acq func. to use)
[GPFlowOPT](https://github.com/GPflow/GPflowOpt) (separate acq, bo object
[Oxford Control](https://github.com/oxfordcontrol/Bayesian-Optimization) (Hierarchical approach, e.g. BO is a subclass of GP, and acq func. is a subclass of BO
and some other.

I feel like to algin with Pyro minimalistic design we should define Acq func. as nn.Modules, and an BO object with simple reference design, and allow user to modify the actually optimization detail by extending the object.
Please let me know what do you think.

**Comment by fehiepsi** [comment]
@tobyclh I think you can make Acquisition a function or a class with `__call__` method (if you need to define some helper methods) as long as there is no parameter needed to optimize. If there are parameters (to be learned), then it is good to define as nn.Module. For the time being, I think a BO object with simple design is good. I guess it will: handle constraints/priors on variables, minimize its acquisition, fit its model?

**Comment by tobyclh** [comment]
@fehiepsi 
I think that will be the way to go.

And for Oxford's approach, the reason they choose to do so is because some acquisition functions when used in batch, requires modifying the optimization procedure itself, a good example is their [BLCB](https://github.com/oxfordcontrol/Bayesian-Optimization/blob/1f9b5ac9f72a8d007caa89dc43fd71a1ec6c484f/methods/blcb.py). If we choose to define BO and acquisition functions separately we might need to find a way to workaround this.

**Comment by eb8680** [comment]
It seems like there's a lot of enthusiasm for this, which is great!  Pyro is a community-driven project, and I'm excited to get #893 merged after it's cleaned up, documented and tested.  

That said, I want to inject a bit of skepticism in this discussion  because there are several mature, high-performance general-purpose GP-based Bayesian optimization libraries and `pyro.contrib.gp` is still missing several crucial BO-related features.  

We should take a step back and ask why it's worth building another BO library on top of Pyro, and what unique value Pyro would add to such a library.  As I see it, if we're serious about doing this, we should aim to produce something that:
1. Exposes a simple interface for auto-tuning of Pyro inference algorithms, and significantly improves upon random search and matches or exceeds other BO libraries in both wall clock time and number of evaluations at the task of hyperparameter selection on several nontrivial Pyro examples (e.g. AIR, DMM, SSVAE)
2. In addition to GP-based BO built on top of `pyro.contrib.gp`, supports Thompson sampling and other Monte Carlo acquisition functions with arbitrary Pyro programs as models and posterior inference with any Pyro inference algorithm
3. Supports high-dimensional compound configuration spaces built from nested data structures containing `torch.distributions.constraints`

I would argue that before investing too much effort in BO, we should write a design doc with a convincing plan to achieve those three points.  An example of a library with similar design goals that was successfully applied to computer system auto-tuning is described in [this paper](https://pdfs.semanticscholar.org/d76b/ebd684d75c39cc9ae5d9033d8fbe415e6845.pdf) and [this dissertation](http://www.cl.cam.ac.uk/techreports/UCAM-CL-TR-900.pdf).  I'll try to post some more detailed thoughts on how a much simpler version of that might look early next week.

**Comment by fehiepsi** [comment]
Agree! In my own opinion, GP is easier to maintain than BO because the technical points are just Math. For BO, there are many situations to cover. So writing a design doc for BO is a very good point to make it in scope. Just sometimes, a developer needs not to be an expert to understand what design doc points out. So it would be helpful to include a step-by-step guide in it if someone starts to write. In addition, I also agree that it is better to have some references to follow as @eb8680 suggested.

**Comment by tobyclh** [comment]
I would argue that there are many BO repositories outside. For example, those repositories I posted either didn't support autograd, or requires very heavy dependency like Tensorflow which uses static graph, which is primarily the reason why I started using pyro and I am sure there is demand for function like this.
But at the same time its very true that BO module is an ambitious project that we should have a good plan if we are going to do it at all. I will read the doc and see how it should be done at my spare time.

**Comment by mhangaard** [comment]
> @ysaatchi @tobyclh It seems to me that computing derivatives is quite straightforward (just make the input variables requires_grad when we need to compute). Let me some time to catch up with literature before I can make any claim.
> 
> For more features of GP, I made a tracking list at #832. Will start to implement them after duo averaging stuff.

Derivatives of observations as in E. Solak, R. Murray-Smith, W. E. Leithead, D. J. Leith, and C. E. Rasmussen,
“Derivative observations in Gaussian process models of dynamic systems,”
in Advances in Neural Information Processing Systems, edited by S. Becker,
S. Thrun, and K. Obermayer (MIT Press, 2003), Vol. 15, pp. 1057–1064.
32 J. Riihimäki and A. Vehtari, “Gaussian processes with monotonicity infor-
mation,” in Proceedings of Machine Learning Research, edited by Y. W. Teh
and M. Titterington (PMLR, 2010), Vol. 9, pp. 645–652.

**Comment by mz1482** [comment]
Is there example for multidimensional bayesian optimization with pyro?

**Comment by fehiepsi** [comment]
@mz1482 For an example, I think the [Bayesian Optimization tutorial](http://pyro.ai/examples/bo.html) works for multidimensional data, with a few changes:
+ change `input_dim=1` of the kernel to the number of dimensions of your data.
+ if your data is constrained (the tutorial assumes the data is constrained in the `interval(0, 1)` but it is not necessary), write extra functions to transform data from constrained domain to unconstrained domain, and reversely.

**Comment by mz1482** [comment]
@fehiepsi Thank you. Yeah that could work. We are having a particular issue where the input space is 3d location (xyz coordinates). It is not even continuous because we have discrete point cloud of space and we want to move to the optimum location. So, in bayesian optimization when Acquisition function will suggest a point, I also have to add additional thing to get the nearest point of the suggested. Because for every point cloud, I get output. Other than any location, we do not have any output.  Thanks again for your answer.
